sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala

```diff
@@ -37,10 +37,13 @@ import org.apache.spark.sql.vectorized.ColumnarBatch
  *                        node during canonicalization.
  * @param partitionSpecs  The partition specs that defines the arrangement, requires at least one
  *                        partition.
+ * @param numEmptyPartitions Number of empty partitions. This is only present when there is
+ *                           coalescing
  */
 case class AQEShuffleReadExec private(
     child: SparkPlan,
-    partitionSpecs: Seq[ShufflePartitionSpec]) extends UnaryExecNode {
+    partitionSpecs: Seq[ShufflePartitionSpec],
+    numEmptyPartitions: Option[Int] = None) extends UnaryExecNode {
```

Comment:
nit: make this `numEmptyPartitions: Int = 0`?

Reply:
Do you think we need to distinguish `None`(for non-coalesce `AQEShuffleReadExec`) and `0`?